core/pathresolver.go

```diff
@@ -55,3 +56,31 @@ func Resolve(ctx context.Context, n *IpfsNode, p path.Path) (*merkledag.Node, er
 	// ok, we have an ipfs path now (or what we'll treat as one)
 	return n.Resolver.ResolvePath(ctx, p)
 }
+
+// ResolveToKey resolves a path to a key.
+//
+// It first checks if the path is already in the form of just a key (<key> or
+// /ipfs/<key>) and returns immediately if so. Otherwise, it falls back onto
+// Resolve to perform resolution of the dagnode being referenced.
+func ResolveToKey(ctx context.Context, n *IpfsNode, p path.Path) (key.Key, error) {
+
+	// If the path is simply a key, parse and return it. Parsed paths are already
+	// normalized (read: prepended with /ipfs/ if needed), so segment[1] should
+	// always be the key.
+	if p.IsJustAKey() {
+		return key.B58KeyDecode(p.Segments()[1]), nil
+	}
+
+	// Fall back onto regular dagnode resolution.
+	dagnode, err := Resolve(ctx, n, p)
+	if err != nil {
+		return key.Key(""), err
+	}
```

Comment:
i mean, suppose we have:

```
/ipfs/<hash>/foo/bar
```
We only need the objects:
```
/ipfs/<hash>
/ipfs/<hash>/foo
```
to complete this request, because we can get the hash of `/ipfs/<hash>/foo/bar` by looking into `/ipfs/<hash>/foo` and grabbing the key there.

so `ResolveToKey` should only need the second to last object, and grab the key from there.